helper/raftutil/state.go

```diff
@@ -160,3 +160,39 @@ func commandName(mt structs.MessageType) string {
 
 	return fmt.Sprintf("%v", mt)
 }
+
+// Find raft.db given path p
+func FindRaftInPath(p string, includeFileInPath bool) (raftpath string, err error) {
+	if _, err := os.Stat(p); err == nil && filepath.Ext(p) == ".db" {
+		// Path p is to a valid .db file, use as specified
+		raftpath = p
+	} else {
+		// Define walk function to identify raft.db
+		walkFn := func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+
+			if filepath.Base(path) == "raft.db" {
+				raftpath = path
+			}
```

Comment:
This is fine.  I considered using a walking approach but was concerned about some edge cases: if the data-dir belongs to a server+client, then it will scan the alloc dirs as well and the walking may fail with permission errors when the alloc dirs are owned by root but invoker is run as unprivileged process.  We can merge this and change it if it's a problem.

Reply:
I refactored the functions to make their usage clear, and also check the hotpaths first to avoid walking the directory if possible.